text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+        );
+
+        Self::validate_layout(new_layout)?;
+
+        let pointer = self.0.get() as *mut u8;
+
+        //  Safety:
+        //  -   Both starting and resulting pointers are in bounds of the same allocated objects as `old_layout` fits
+        //      `pointer`, as per the pre-conditions of `grow_zeroed`.
+        //  -   The offset does not overflow `isize` as `old_layout.size()` does not.
+        let pointer = unsafe { pointer.add(old_layout.size()) };
+
+        //  Safety:
+        //  -   `pointer` is valid, since `self` is valid.
+        //  -   `pointer` points to at an area of at least `mem::size_of::<T>() - old_layout.size()`.
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StableStore for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
+
+    /// Return a pointer to the block of memory associated to `handle`.
+    unsafe fn resolve(&self, handle: Self::Handle) -> NonNull<u8>;
+
+    //  The following methods are similar to `Allocator`, reformulated in terms of `Self::Handle`.
+
+    /// Allocates a new handle to a block of memory. On success, invalidates any existing handle.
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError>;
```

Comment:
cc the conversations about whether `Layout` is the best option, or if there should be a `NonZstLayout` or something instead <https://rust-lang.zulipchat.com/#narrow/stream/197181-t-libs.2Fwg-allocators/topic/Forbidding.20Zero-sized.20layouts.20in.20Allocator.3A.3Aallocate.2E/near/365187429>

Not sure if this RFC is the place to resolve that, though.

Reply:
Yes, is likely to have the same issues described in <https://github.com/rust-lang/rust/issues/32838#issuecomment-1585824292> and <https://github.com/rust-lang/rust/issues/32838#issuecomment-1585940909>.

Additionally, it's probably worth including the tweak to growth I mentioned here: <https://rust-lang.zulipchat.com/#narrow/stream/197181-t-libs.2Fwg-allocators/topic/Allocator.20growth.20method.20and.20length.20param>

I don't expect this to significantly change the shape of things, though.

Reply:
I think those discussions are somewhat orthogonal to the general idea of this RFC, indeed.

I'm happy to follow whatever consensus emerge, be it on the `NonZstLayout`, tweaks to `grow`, etc...

Reply:
Putting `dangling` on the trait gives the interesting option of that being a method for ZST allocation. It's directly relevant to the RFC in as much as there needs to be a way to allocate ZSTs in storage.

`dangling` giving handles that are valid to resolve also gives it a more concrete reason to take `&self`.

Reply:
(For what it's worth, given the entire rest of the interface needs to be `unsafe`, I'd have a mild preference towards just making `allocate` unsafe and the use of a non-zero-sized layout for allocation a safety requirement, despite being the person who pushed `size <= isize::MAX` through to be included in `Layout`'s guarantees. But that std got it wrong (leaky) with `Vec`, that acts as similar evidence that consistently getting it right without compiler assistance is problematically difficult.)

I think the interface I'd most like to _use_ is a `const fn allocate_dangling(&self, Layout) -> Result<Self::Handle, AllocError>` which is guaranteed to not do any real allocation work (making deallocation optional) but still participates in the usual handle (in)validation and can be resolved, reallocated, and/or deallocated. The other ways of allocation get a weaker guarantee; allocation of a zero-byte handle may consume resources from the storage, but forgetting to deallocate a zero-byte handle must not result in leaking any more memory than when deallocating it once the storage is dropped _or_ forgotten (i.e. wasting space in local or borrowed storage is acceptable, but static storage not). This is an annoying spec to provide, but I think captures the desire of using such an interface.

Reply:
Prompted by @scottmcm I experimented with making `Store::dangling` return a handle to a well-aligned block of memory, though apart from being resolvable (to an invalid pointer) the handle remains unusable with deallocate/grow/shrink. You can see the experiment at https://github.com/matthieu-m/storage/tree/experiment/dereferenceable-dangling-handle .

It's somewhat similar to `allocate_dangling`:

- It takes the new `ptr::Alignment` type as argument, rather than a full-blown `Layout`, since the size is unnecessary (and possibly misleading).
- It becomes fallible.

I'd favor concentrating all discussions on dangling allocations in a single comment thread, and therefore would recommend pursuing `allocate_dangling` in https://github.com/rust-lang/rfcs/pull/3446#discussion_r1240786356 if you don't mind.

_PS: regarding making `allocate` or `dangling` `unsafe`, I have no a priori rejection, though I do find a bit weird to have `unsafe` on methods with no pre-conditions._

Reply:
`allocate` would only be made `unsafe` if it gained a precondition (e.g. a nonzero allocation size), yes.